Locally Predictive LDA

charleslparker opened the pull request:
Here it is, @mmerce .  I tried to follow conventions as best I could, but I'm sure I've missed some things, especially documentation-wise.  I've also never seen the tests structured as they are here, and I again tried to match but may have failed.

In any case, you should get the general idea from the tests.  Let me know what needs fixing.  Thanks!

/cc @jaor in case he wants to comment.

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+}
+
+class LDA(ModelFields):
+    """ A lightweight wrapper around a cluster model.
```
.... around an LDA model

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+    """ A lightweight wrapper around a cluster model.
+
+    Uses a BigML remote cluster model to build a local version that can be used
+    to generate centroid predictions locally.
```
ditto: changing cluster to LDA and centroid to distribution or topic distribution

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+
+"""
+from bigml.api import FINISHED
+from bigml.api import (BigML, get_lda_id, get_status)
```
I believe parethesis are only needed when you need multiple lines

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+            self.resource_id = get_lda_id(lda_model)
+            if self.resource_id is None:
+                raise Exception(api.error_message(lda_model,
+                                                  resource_type='cluster',
```
resource_type='lda'

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+        # Checks and cleans input_data leaving the fields used in the model
+        input_data = self.filter_input_data(input_data, by_name=by_name)
+
+        # Checks that all numeric fields are present in input data
```
I'm not sure about this code. From the error message I assume that LDA needs all the text fields to be provided in the input data. The comment is referring to numeric fields (I guess it's inherited from the cluster code) and the real code checks all the available fields. If you build an LDA model on a dataset with numeric and categorical fields in addition to the text ones, will they appear in this "fields" structure?

charleslparker replied:
> From the error message I assume that LDA needs all the text fields to be provided in the input data.

Right - the comment is wrong, I'm just trying to make sure all of the fields in the model array are present in the input data.  Am I doing it correctly?

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+
+    def _stem(self, term):
+        if not self._stemmer:
+            return turn
```
return term?

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,310 @@
+
+                i += 1
+
+        if (self._bigrams and
```
If I was jao, I'd ask you to build a function to do this segment of code, whiich is repeated in the while loop and here...

mmerce commented on bigml/tests/compute_lda_prediction_steps.py:
```diff
@@ -0,0 +1,39 @@
+
+    all_close = True
+
+    if len(distribution) != len(expected):
```
I believe you can improve this code with 

http://nullege.com/codes/search/nose.tools.assert_almost_equals

http://nose.readthedocs.io/en/latest/testing_tools.html

using eq_ and assert_almost_equals

mmerce commented on bigml/tests/test_32_lda_prediction.py:
```diff
@@ -0,0 +1,96 @@
+            ]
+        ]
+
+        for ex in examples:
```
This test is fine to ensure the function works when used with the given model, but imagine the lda structure returned by  the API changes. Then you'll never notice. We usually build first a model using a toy dataset. Thus we can also use this test to cross validate the API. If you have a look at test_05_compare_predictions, you'll see that we create a remote prediction and a local one to check whether they are equal.

charleslparker replied:
> This test is fine to ensure the function works when used with the given model, but imagine the lda structure returned by the API changes.

Right; we'll need another test to test whatever is returned from the server.  I only have this test in there because I want to have an exact copy of the test I have in the `bigmlcom/streaming-lda` library, for my own sanity.

I would write the test against the server, but I didn't know if I could do it at this stage of deployment.  Am I able to train toy LDA models against production now?

mmerce replied:
I don't think it's available, not even in development machines, so we'll have to wait to add this test. No problem.

mmerce commented:
Two general comments. It's fine if you want to use private attibutes, but then it's a little bit strange to me that you set as private some attributes that are exposed as public in the JSON response, like bigrams, case_sensitive, alpha... they are not to be modified but they are public, right?

And the second comment is that I never shorten the names of the variables, even if they are indexes in a for loop. Francisco's seminal code never did that, and I followed his convention for clarity. It might be tiresome sometimes to write so much, but is kind of helpful sometimes for others to understand. Even more when you don't comment your private functions. I'm not sure pep8 or pylint won't complain about that also...

Apart from that, everything looks good. If you want to keep this unit test and let me add another one to compare local and remote predictions, it's also fine.

charleslparker commented:
I've addressed most of the comments, I think.

> It's fine if you want to use private attibutes, but then it's a little bit strange to me that you set as private some attributes that are exposed as public in the JSON response, like bigrams, case_sensitive, alpha... they are not to be modified but they are public, right?

I've always thought of privacy mainly as a signal to subclassers about what you can touch/override without breaking the superclass's functionality.  But I looked in cluster.py and it looks like you've got nothing private in there, so I've removed all privacy for consistency with that.  Let me know if I should do something different.

> I never shorten the names of the variables, even if they are indexes in a for loop.

I've tried to expand as many variable names as possible; the only ones I've kept short were cases where 1.) I'm directly following names in the research paper that outlines the implementation I'm using or 2.) I have code that uses those exact names elsewhere.  If there are particular ones you think need expanding for clarity, do let me know. 

I've also commented all functions, so have a quick look please at that to make sure that my English is characteristically sloppy.

Thanks!

mmerce commented on bigml/lda.py:
```diff
@@ -0,0 +1,337 @@
+                sums = [sum(n[index] for n in assignments) for index
+                        in range(self.ntopics)]
+
+                self.phi = [[0 for _ in range(nterms)]
```
nit: isn't it like?

```python
    [[0] * len(nterms)] * len(self.ntopics)
```

mmerce commented:
Just a nit that you can ignore. Running the tests.
